### ISSUE TITLE
Allow dynamic subscription for Java client

### ISSUE THREAD

**Issue by Iteratrix** [cometd/cometd#798]
[CometD already clearly documents](https://docs.cometd.org/current4/reference/#_javascript_subscribe_resubscribe) how to handle dynamic subscription and unsubscription after a handshake has completed for the Javascript client. However, I haven't been able to find any similar documentation for the Java client, and the technique described for Javascript doesn't seem to translate to the Java library.

If there's an existing solution for dynamic subscribing/unsubscribing in the Java client, that should be documented. If not, it'd be nice to include that feature in a future release.

Thanks.

**Comment by sbordet** [comment]
In the CometD JavaScript library, `cometd.subscribe(...)` returns an opaque data structure that must be passed to `cometd.unsubscribe(...)`.
In the CometD Java library, `ClientSessionChannel.subscribe(...)` takes a `MessageListener` and returns a `boolean`, so there is no opaque data structure to be passed to unsubscribe - in fact the correspondent data structure is the `MessageListener` instance.
Because of the absence of this opaque data structure in the Java library, the `MessageListener` needs to be stored (e.g. as a field) to eventually pass it to `unsubscribe(...)`.

A dynamic resubscription in case of re-handshake would be, in Java:

```java
class Controller {
  private final BayeuxClient _cometdClient;
  private MessageListener _subscription;

  public Controller(BayeuxClient cometdClient) {
    _cometdClient = cometdClient;
  }
  
  public void dynamicSubscribe() {
    _subscription = this::onEvent;
    _cometdClient.getChannel("/dynamic").subscribe(_subscription);
  }

  public void dynamicUnsubscribe() {
    if (_subscription != null) {
      _cometdClient.getChannel("/dynamic").unsubscribe(_subscription);
      _subscription = null;
    }
  }
  
  public void onEvent(ClientSessionChannel channel, Message message) {
    ...
  }

  public void handshake() {
    cometdClient.handshake(message -> {
      if (message.isSuccessful()) {
        if (_subscription != null) {
          _cometdClient.getChannel("/dynamic").subscribe(_subscription);
        }
    });
  }
}
```

So basically in the CometD Java library you already have the "resubscribe" functionality you have in the CometD JavaScript library.
